tools/circt-lec/Solver.cpp

```diff
@@ -0,0 +1,193 @@
+//===-- Solver.h - SMT solver interface -------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// This file defines a SMT solver interface for the 'circt-lec' tool.
+///
+//===----------------------------------------------------------------------===//
+
+#include "Solver.h"
+#include "Circuit.h"
+#include "LogicExporter.h"
+#include "Utility.h"
+#include "mlir/IR/Builders.h"
+#include <string>
+#include <z3++.h>
+
+#define DEBUG_TYPE "solver"
+
+Solver::Solver(mlir::MLIRContext *mlirCtx, bool statisticsOpt)
+    : circuits{}, mlirCtx(mlirCtx), context(), solver(context) {
+  this->statisticsOpt = statisticsOpt;
+}
+
+Solver::~Solver() {
+  delete circuits[0];
+  delete circuits[1];
+}
+
+/// Solve the equivalence problem between the two circuits, then present the
+/// results to the user.
+mlir::LogicalResult Solver::solve() {
+  // Constrain the circuits for equivalence checking to be made:
+  // require them to produce different outputs starting from the same inputs.
+  if (constrainCircuits().failed())
+    return mlir::failure();
+
+  // Instruct the logical engine to solve the constraints:
+  // if they can't be satisfied it must mean the two circuits are functionally
+  // equivalent. Otherwise, print a model to act as a counterexample.
+  mlir::LogicalResult outcome = mlir::success();
+  switch (solver.check()) {
+  case z3::unsat:
+    lec::outs() << "c1 == c2\n";
+    break;
+  case z3::sat:
+    lec::outs() << "c1 != c2\n";
+    printModel();
+    outcome = mlir::failure();
+    break;
+  case z3::unknown:
+    outcome = mlir::failure();
+    lec::errs() << "circt-lec error: solver ran out of time\n";
+  }
+
+  // Print further relevant information as requested.
+  LLVM_DEBUG(printAssertions());
+  if (statisticsOpt)
+    printStatistics();
+
+  return outcome;
+}
+
+/// Create a new circuit to be compared and return it.
+Solver::Circuit *Solver::addCircuit(llvm::StringRef name, bool firstCircuit) {
+  // Hack: entities within the logical engine are namespaced by the circuit
+  // they belong to, which may cause shadowing when parsing two files with a
+  // similar module naming scheme.
+  // To avoid that, they're differentiated by a prefix.
+  std::string prefix = firstCircuit ? "c1@" : "c2@";
+  circuits.push_back(new Solver::Circuit(prefix + name, this));
+  assert(circuits.size() <= 2 && "expected to solve two circuits"); // NOLINT
```

Comment:
I would argue it's an assertion because it couldn't be triggered from user input. The LLVM Coding Standards manual recommends asserting preconditions and provides examples where they're used for bound checking.